include/deal.II/grid/manifold_lib.h

```diff
@@ -503,6 +504,182 @@ class TorusManifold : public ChartManifold<dim,3,3>
+ * of the computational domain. The outer curved boundary description is assumed to
+ * be given by another manifold (e.g. a polar manifold on a circle). The mechanism to
+ * extend the boundary information is a so-called transfinite interpolation.
+ *
+ * The formula for extending such a description in in 2D is inspired by
+ * <a href="https://en.wikipedia.org/wiki/Transfinite_interpolation">
+ * Wikipedia</a>
+ * Given a point $(u,v)$ on the chart, the image of this point in real space
+ * is given by
+ * @f{align*}{
+ * \bf S(u,v) &= (1-v)\bf c_0(u)+v \bf c_1(u) + (1-u)\bf c_2(v) + u \bf c_3(v) \\
+ * &\quad - \begin[(1-u)(1-v) \bf x_0 + u(1-v) \bf x_1 + (1-u)v \bf x_2 + uv \bf x_3 \right]
+ * @f}
+ * where $\bf x_0, \bf x_1, \bf x_2, \bf x_3$ denote the four bounding vertices
+ * bounding the image space and $\bf c_0, \bf c_1, \bf c_2, \bf c_3$ are the
+ * four curves describing the lines of the cell. If a curved manifold is
+ * attached to any of these lines, the evaluation is done according to
+ * Manifold::get_new_point() with the two end points of the line and
+ * appropriate weight. In 3D, the generalization of this formula is
+ * implemented, creating a weighted sum of the vertices (positive
+ * contribution), the lines (negative), and the faces (positive contribution).
+ *
+ * This manifold is usually attached to a coarse mesh and then places new
+ * points as a combination of the descriptions on the boundaries, weighted
+ * appropriately according to the position of the point on the original chart
+ * point $(u,v). Whenever possible, this manifold should be preferred over
+ * setting only a curved manifold on the boundary of a mesh, since the latter
+ * will need to switch from a curved description to a straight description in a
+ * single layer of elements, which causes an error order on cells close to the
+ * boundary that does not exceed 3 no matter how high the degree of the
+ * polynomial mapping and the finite element space. Using this class instead,
+ * the curved nature of the manifold that is originally contained in one
+ * <i>coarse</i> mesh layer will be applied to more than one <i>fine</i> mesh
+ * layer once the mesh gets refined, restoring the optimal convergence rates of
+ * the underlying finite element and mapping as appropriate.
+ *
+ *
+ * @author Luca Heltai, Martin Kronbichler, 2017
+ */
+template <int dim, int spacedim=dim>
+class TransfiniteInterpolationManifold : public Manifold<dim,spacedim>
+{
+public:
+  /**
+   * Constructor.
+   */
+  TransfiniteInterpolationManifold();
+
+  /**
+   * Initializes the manifold with a coarse mesh. The prerequisite for using
+   * this class is that the input triangulation is uniformly refined and the
+   * manifold is later attached to the same triangulation.
+   *
+   * @note For performance and accuracy reasons, it is recommended to apply the
+   * transfinite manifold to as coarse a mesh as possible. Regarding accurcay,
```

Comment:
I have a problem with this recommendation: Looking at cylindrical part of the figure (from #4451) towards the middle:
![grid_cylinder](https://cloud.githubusercontent.com/assets/8276200/26629673/b6b12202-4603-11e7-9d27-04c6bee17b5d.png)
shows that some cells are strange/inverted. This is because the detection of the coarse mesh cell fails: Two points along the circular shape end up outside the domain spanned by `MappingQ1` that I use for the identification of the coarse cell, and they get closer to the neighbor. It can be fixed by going to a mesh one level finer:
![grid_cylinder_2](https://cloud.githubusercontent.com/assets/8276200/26630576/a12d04c4-4607-11e7-8880-428086fc4d25.png)

Thus, we should probably recommend to try as coarse mesh as possible but if one detects problems initialize again on a mesh one level finer.

It would of course have been even better if we weren't throwing away the context of the current mesh cell in `get_new_point` or `add_new_points` by passing the cell where we actually want to compute the new points. This avoids doing a manual search which might get things wrong since I'd only need to go to the appropriate parent. But changing interface because of this one class is not what I want to do.